src/rules/noConstantBinaryExpressionRule.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isBinaryExpression } from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-constant-binary-expression",
+        description: "Both sides of a binary expression can't be the same.",
+        optionsDescription: "Not configurable.",
+        options: null,
+        optionExamples: [true],
+        type: "typescript",
+        typescriptOnly: true,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING = "Both sides of a binary expression can't be the same";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk);
+    }
+}
+
+function walk(ctx: Lint.WalkContext<void>) {
+    return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if (isBinaryExpression(node)) {
```

Comment:
you need to be careful here. BinaryExpressions is everything where you have `<expression> <operator> <expression>`.
you need to check the kind of the operator to avoid errors on `foo + foo`, `foo -= foo`, `foo(), foo()` and probably a lot more.

If this rule is only intended for comparisons your are interested in `EqualsEqualsToken`, `EqualsEqualsEqualsToken`, `ExclamationEqualsToken`, `ExclamationEqualsEqualsToken`.
If you also want to show an error on `a = a`, you need `EqualsToken` as well

Reply:
I would love to have a rule for a=a. Should it be in the same rule? This is an assignment, not a comparaison, the rule should have a different name so? Or have a different rule? `no-self-assignement`?

Reply:
Makes sense to split this into two separate rules unless someone can come up with a name that covers both cases.

Reply:
Not to myself: this should error on `a < a` and so on (`a<=a, a>a, a>=a`)